In [97]:
import os
import pandas as pd
import numpy as np
import gzip
import ijson
from ast import literal_eval
import networkx as nx
import itertools
import re

## Directories

In [2]:
DATA_DIR = os.getenv("DATA_DIR")
PREDICT_DIR = os.path.join(DATA_DIR, "predict_network")
content_api = os.path.join(DATA_DIR, "content_api")
sub_content_api = os.path.join(content_api,"all_dates")
# api_links_file = os.path.join(content_api,"07-02-19", "content_api_links.csv.gz")
labelled_file = os.path.join(content_api,"labelled.csv.gz")

## Initialized labelled/content

In [3]:
labelled = pd.read_csv(labelled_file, compression="gzip", usecols=['base_path', 'content_id', 'title'])

In [4]:
labelled.drop_duplicates("content_id", inplace=True)

In [5]:
len(labelled.base_path)

206038

In [6]:
labelled[labelled.base_path=="/sign-in-universal-credit"]

base_path                            content_id  \
272249  /sign-in-universal-credit  b220a437-0d51-4390-9993-63345d0c83ad   

                                           title  
272249  sign in to your universal credit account

## Read in links data

In [187]:
api_links_files = [os.path.join(sub_content_api, f) for f in os.listdir(sub_content_api) if "api_links" in f]
api_links_files

['/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/missing_urls_240219_content_api_links.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/labelled2_content_api_links.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/pt01of9_content_api_links.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/content_api_links07.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/content_api_links_remote.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/content_api_links_22.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/labelled1_content_api_links.csv.gz',
 '/Users/felisialoukou/Documents/govuk-network-embedding/data/content_api/all_dates/pt02of9_content_api_links.csv.gz',
 '/Users/felisialoukou/Documents/govuk-netwo

In [188]:
links = pd.concat([pd.read_csv(f, compression="gzip") for f in api_links_files], ignore_index=True)
links.drop_duplicates("url", inplace=True)
links.shape, links.columns

((231145, 7),
 Index(['url', 'embedded_links', 'related_links', 'collection_links', 'num_rel',
        'num_emb', 'num_coll'],
       dtype='object'))

In [189]:
missing_urls = set(labelled.base_path).difference(set(links.url))
print(len(missing_urls))

4


In [190]:
with gzip.open(os.path.join(sub_content_api, "missing_urls_250219.csv.gz"), "wt") as w:
    w.write("Node\n")
    for missing in missing_urls:
        w.write("{}\n".format(missing))

In [191]:
link_types = ['embedded_links',  'collection_links', 'related_links']

def lit_eval(x):
    return [l.strip() for l in re.sub("\[|\]|'","",x).split(",")]

for ltype in link_types:
    print(ltype)
    links[ltype] = links[ltype].map(lambda x: lit_eval(x))

embedded_links
collection_links
related_links


In [192]:
links[links.url=="/sign-in-universal-credit"].related_links.values[0][1]

'/apply-universal-credit'

In [193]:
links.columns.get_loc("embedded_links")

1

In [194]:
adj_list = []
for tup in links.itertuples(index=False):
    for ltype in link_types:
        tup_ind = links.columns.get_loc(ltype)
        if len(tup[tup_ind]) > 0:
            for dest in tup[tup_ind]:
#                 if dest in labelled.base_path.values:
                adj_list.append((tup.url, dest, ltype))   

In [195]:
# ((231145, 7), 887717)
links.shape, len(adj_list)

((231145, 7), 887717)

In [196]:
graph = nx.DiGraph()

In [197]:
for s,d,t in adj_list:
    graph.add_edge(s,d,link_type=t)    

In [198]:
nx.info(graph)

'Name: \nType: DiGraph\nNumber of nodes: 437808\nNumber of edges: 512879\nAverage in degree:   1.1715\nAverage out degree:   1.1715'

In [199]:
base_cid = dict(zip(labelled.base_path, labelled.content_id))

In [200]:
rowlist = []
for s,d,t in adj_list:
    row = {'src_node':s,
    'dest_node': d,
    'link_type': t,
    'src_cid': base_cid[s] if s in base_cid.keys() else np.nan,
    'dest_cid': base_cid[d] if d in base_cid.keys() else np.nan}
    rowlist.append(row)

In [201]:
df_structural = pd.DataFrame(rowlist)

In [202]:
df_structural.shape, df_structural.dropna(subset=["dest_cid","src_cid"]).shape

((887717, 5), (62660, 5))

In [203]:
structural_file = os.path.join(content_api, "structural_training_data_no_node_id.csv.gz")
df_struct_filter = df_structural.dropna(subset=["dest_cid","src_cid"]).copy(deep=True)
df_struct_filter.to_csv(structural_file, index=False, compression="gzip")

## Merge structural with functional data

In [204]:
edgefile = os.path.join(DATA_DIR, "processed_network", "edges_graphsagetest_feb_01_18_doo_min15weight_wtext.csv.gz")
edges = pd.read_csv(edgefile, compression="gzip")
edges.head()

src_node  \
0  /government/publications/guidance-for-dependan...   
1                                         /visa-fees   
2                  /entering-staying-uk/family-visas   
3                                    /uk-family-visa   
4                     /uk-family-visa/partner-spouse   

                                           dest_node  weight  \
0                                         /visa-fees      66   
1                    /find-a-visa-application-centre    1164   
2                                    /uk-family-visa     377   
3                     /uk-family-visa/partner-spouse   26227   
4  /government/publications/application-for-uk-vi...    2733   

                                src_cid                              dest_cid  \
0  5ef7560d-7631-11e4-a3cb-005056011aef  1e333395-5dd5-4452-96a3-fbe939928761   
1  1e333395-5dd5-4452-96a3-fbe939928761  aa055bd6-fde1-4471-8ac2-ac86241b4898   
2  d612c61e-22f4-4922-8bb2-b04b9202126e  305acc88-488e-44ec-8e5a-1a5de7819ba9   
3  305acc88-488e-44ec-8e5a-1a5de7819ba9  305acc88-488e-44ec-8e5a-1a5de7819ba9   
4  305acc88-488e-44ec-8e5a-1a5de7819ba9  5ef421be-7631-11e4-a3cb-005056011aef   

   source  target  
0   33265   58314  
1   58314   12343  
2   23720   25053  
3   25053   14044  
4   14044   33797

In [205]:
node_id = dict(zip(edges.src_node, edges.source))
for node,nid in zip(edges.dest_node, edges.target):
    if node not in node_id.keys():
        node_id[node] = nid

In [206]:
df_struct_filter['source'] = df_struct_filter['src_node'].map(lambda x: node_id[x] if x in node_id.keys() else -1)
df_struct_filter['target'] = df_struct_filter['dest_node'].map(lambda x: node_id[x] if x in node_id.keys() else -1)

In [207]:
df_struct_filter.head(2)

dest_cid  \
194  5ef78617-7631-11e4-a3cb-005056011aef   
195  5ef78888-7631-11e4-a3cb-005056011aef   

                                             dest_node         link_type  \
194  /government/publications/transxchange-examples...  collection_links   
195     /government/publications/transxchange-overview  collection_links   

                                  src_cid  \
194  5ef7544d-7631-11e4-a3cb-005056011aef   
195  5ef7544d-7631-11e4-a3cb-005056011aef   

                                 src_node  source  target  
194  /government/collections/transxchange      -1      -1  
195  /government/collections/transxchange      -1      -1

In [208]:
df_struct_filter.shape

(62660, 7)

In [209]:
# df_struct_filter[df_struct_filter.source==-1].src_node
# df_struct_filter[df_struct_filter.target==-1].dest_node
# df_struct_filter.dest_node
def compute_coverage(src_node, dest_node, miss_src, miss_dest):
    ### Number of unique nodes
    unique_nodes = set(list(src_node) + list(dest_node))
    print(len(unique_nodes))
    missing_nodes = list(set(miss_src))
    missing_nodes.extend(list(set(miss_dest)))
    print(len(missing_nodes))
    print("\n {} nodes in funct, but not in struct".format(len(set(node_id.keys()).difference(unique_nodes))),"\n",
    "{} nodes both in funct and struct data".format(len(set(node_id.keys()).intersection(unique_nodes))),"\n",
    "{} nodes in struct, but not in funct".format(len(set(unique_nodes).difference(node_id.keys()))))

In [210]:
compute_coverage(df_struct_filter.src_node, 
                 df_struct_filter.dest_node, 
                 df_struct_filter[df_struct_filter.source==-1].src_node, 
                 df_struct_filter[df_struct_filter.target==-1].dest_node)

52323
41416

 55541 nodes in funct, but not in struct 
 11078 nodes both in funct and struct data 
 41245 nodes in struct, but not in funct


     55541 nodes in funct, but not in struct 
     11078 nodes both in funct and struct data 
     41245 nodes in struct, but not in funct

In [211]:
max_node = max(node_id.values())+1
for node in missing_nodes:
    if node not in node_id.keys():
        node_id[node] = max_node
        max_node+=1

In [212]:
df_struct_filter['source'] = df_struct_filter['src_node'].map(lambda x: node_id[x] if x in node_id.keys() else -1)
df_struct_filter['target'] = df_struct_filter['dest_node'].map(lambda x: node_id[x] if x in node_id.keys() else -1)

In [214]:
compute_coverage(df_struct_filter.src_node, 
                 df_struct_filter.dest_node, 
                 df_struct_filter[df_struct_filter.source==-1].src_node, 
                 df_struct_filter[df_struct_filter.target==-1].dest_node)

52323
0

 55541 nodes in funct, but not in struct 
 52323 nodes both in funct and struct data 
 0 nodes in struct, but not in funct


In [216]:
df_struct_filter.shape

(62660, 7)

In [220]:
merged_edges = pd.concat([edges, df_struct_filter.drop('link_type', axis=1)], ignore_index=True, sort=True)

In [223]:
edges.shape, merged_edges.shape

((213446, 7), (276106, 7))

In [ ]:
### Figure out whicj

In [227]:
merged_edges['edge-id'] = merged_edges[['source','target']].apply(lambda x: "{}-{}".format(x[0],x[1]), axis=1)
print(merged_edges.drop_duplicates("edge-id").shape)
ids = merged_edges['edge-id']
merged_edges[ids.isin(ids[ids.duplicated()])].sort_values("edge-id")

(263756, 8)


dest_cid  \
125396  5f20dfc0-7631-11e4-a3cb-005056011aef   
268771  5f20dfc0-7631-11e4-a3cb-005056011aef   
133206  0f82316f-4fd5-45f9-b992-c14ad7504f25   
265252  0f82316f-4fd5-45f9-b992-c14ad7504f25   
265251  2aa74b54-b119-4dd7-8a75-ddbb030458e8   
200859  2aa74b54-b119-4dd7-8a75-ddbb030458e8   
242475  48c0c30d-5f04-4095-8804-d1ee8989761d   
242474  48c0c30d-5f04-4095-8804-d1ee8989761d   
46092   48c0c30d-5f04-4095-8804-d1ee8989761d   
237727  1d5c1538-8599-4174-8540-8fb14ba3185c   
85025   1d5c1538-8599-4174-8540-8fb14ba3185c   
237736  1d5c1538-8599-4174-8540-8fb14ba3185c   
27027   fdcc0a00-5683-47c4-9e17-9b81bb41d9b7   
237729  fdcc0a00-5683-47c4-9e17-9b81bb41d9b7   
89387   d6865b56-80a8-43c7-a447-fb15b8c12760   
237732  d6865b56-80a8-43c7-a447-fb15b8c12760   
64102   d99d777b-2d47-4764-b05a-b9be25ee1a8f   
237725  d99d777b-2d47-4764-b05a-b9be25ee1a8f   
237731  91504a10-f697-42a3-a779-f238b4955ea9   
23493   91504a10-f697-42a3-a779-f238b4955ea9   
118941  6b98928a-0261-4b7b-bbf8-a538955120a7   
237735  6b98928a-0261-4b7b-bbf8-a538955120a7   
28463   435fb04f-2b9f-4f44-8b41-2a962e8c46a8   
250589  435fb04f-2b9f-4f44-8b41-2a962e8c46a8   
67161   aab159ca-a3ff-4846-bfca-745c699e9373   
250590  aab159ca-a3ff-4846-bfca-745c699e9373   
139083  fd3b5ae3-cc11-4696-9834-912347d5198d   
245541  fd3b5ae3-cc11-4696-9834-912347d5198d   
78481   60242728-7631-11e4-a3cb-005056011aef   
245539  60242728-7631-11e4-a3cb-005056011aef   
...                                      ...   
20060   087edbc6-1cfa-4444-80de-84c888c7de71   
243535  087edbc6-1cfa-4444-80de-84c888c7de71   
243536  88c87490-1678-45a4-8fbb-41f9f256f3b2   
102820  88c87490-1678-45a4-8fbb-41f9f256f3b2   
173694  69c1bb26-aba4-4884-9cba-328b6e10143a   
243528  69c1bb26-aba4-4884-9cba-328b6e10143a   
95353   1190a71a-208c-4f52-b7cd-06a5d371f804   
243539  1190a71a-208c-4f52-b7cd-06a5d371f804   
83000   5cb58486-0b00-4da8-8076-382e474b4f03   
245518  5cb58486-0b00-4da8-8076-382e474b4f03   
58990   602917fc-7631-11e4-a3cb-005056011aef   
245517  602917fc-7631-11e4-a3cb-005056011aef   
66724   0a8423f3-8f91-4769-85a7-f5e985e69009   
245519  0a8423f3-8f91-4769-85a7-f5e985e69009   
243691  f790dc71-386e-4440-9689-31f94e7ac64d   
172653  f790dc71-386e-4440-9689-31f94e7ac64d   
150582  3613e6ff-ef8b-4ab1-bf5c-ef9b875962d2   
243694  3613e6ff-ef8b-4ab1-bf5c-ef9b875962d2   
263512  a25ec77f-82aa-4a7e-b589-873cb06710a1   
212155  a25ec77f-82aa-4a7e-b589-873cb06710a1   
3666    a57d3670-b8bd-4aac-92dc-7dcebd1e55d8   
235436  a57d3670-b8bd-4aac-92dc-7dcebd1e55d8   
61349   8834bc7c-7ac7-4d45-8a57-1d5a5dcd70a0   
235435  8834bc7c-7ac7-4d45-8a57-1d5a5dcd70a0   
136519  47f980aa-ff7b-481a-9a48-535400952115   
235438  47f980aa-ff7b-481a-9a48-535400952115   
235437  fbaa8140-e364-405f-bc09-d25ed8c18838   
31652   fbaa8140-e364-405f-bc09-d25ed8c18838   
48274   9acf51ab-0eb4-4134-8d69-82c77c10e60d   
258494  9acf51ab-0eb4-4134-8d69-82c77c10e60d   

                                                dest_node  source  \
125396  /guidance/comply-with-the-compulsory-beef-labe...   10016   
268771  /guidance/comply-with-the-compulsory-beef-labe...   10016   
133206            /government/publications/visit-guidance   10039   
265252            /government/publications/visit-guidance   10039   
265251  /government/publications/considering-human-rig...   10039   
200859  /government/publications/considering-human-rig...   10039   
242475          /what-different-qualification-levels-mean   10059   
242474          /what-different-qualification-levels-mean   10059   
46092           /what-different-qualification-levels-mean   10059   
237727  /government/publications/application-to-recons...    1008   
85025   /government/publications/application-to-recons...    1008   
237736  /government/publications/application-to-recons...    1008   
27027   /government/publications/policy-paper-on-citiz...    1008   
237729  /government/publications/policy-paper-on-citiz...    1008   
89387                   

In [228]:
merged_edges.drop_duplicates("edge-id", inplace=True)

In [231]:
merged_edges[(merged_edges.src_node=="/universal-credit") & (merged_edges.weight.isna())]

dest_cid  \
234998  0b0ff1cc-1532-4cd3-836b-0b7ab0359a4b   
235003  28774875-4c69-41f0-a9a3-49abb015b34d   
235004  84f68c01-c5f3-4d20-9b31-d46fec04498c   
235006  a5d44e01-f9b1-48e4-87e0-96eba6def07b   
235008  145d29c4-cb5c-4b1c-8664-094912f8a34e   
235009  2e89e098-f5a8-4a6c-9dc9-ebe0c7b02ae6   
235010  5ef648f9-7631-11e4-a3cb-005056011aef   
235017  c1b3a2fe-9bf5-44ee-a76c-7d34d4880023   
235018  4479426d-51c1-4b51-8930-543d994e8818   
235020  73d634c6-e3d8-4810-abb1-61b8a9a70c88   
235021  95544523-5e1e-4754-9d87-d1c846f073c0   
235029  58259953-e6d0-4b20-9922-5313fb23534a   
235030  a65ee7cf-7665-4741-834d-2d840871d279   
235031  7d71d8e3-0b33-4757-b8a7-e97a34ee6497   

                                                dest_node  source  \
234998                                      /call-charges   26857   
235003                                  /funeral-payments   26857   
235004                                         /legal-aid   26857   
235006                           /help-with-prison-visits   26857   
235008                        /disabled-facilities-grants   26857   
235009                         /energy-company-obligation   26857   
235010  /government/publications/claiming-discretionar...   26857   
235017                          /court-fees-what-they-are   26857   
235018                           /apply-free-school-meals   26857   
235020                         /mandatory-reconsideration   26857   
235021                                     /healthy-start   26857   
235029  /government/collections/universal-credit-detai...   26857   
235030  /government/collections/universal-credit-infor...   26857   
235031  /government/collections/universal-credit-infor...   26857   

                                     src_cid           src_node  target  \
234998  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   18430   
235003  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   51325   
235004  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   14677   
235006  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   12935   
235008  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   34166   
235009  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   39394   
235010  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit    2099   
235017  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   50126   
235018  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   54857   
235020  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   60461   
235021  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit     487   
235029  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   45139   
235030  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit   15420   
235031  f790dc71-386e-4440-9689-31f94e7ac64d  /universal-credit    2009   

        weight      edge-id  
234998     NaN  26857-18430  
235003     NaN  26857-51325  
235004     NaN  26857-14677  
235006     NaN  26857-12935  
235008     NaN  26857-34166  
235009     NaN  26857-39394  
235010     NaN   26857-2099  
235017     NaN  26857-50126  
235018     NaN  26857-54857  
235020     NaN  26857-60461  
235021     NaN    26857-487  
235029     NaN  26857-45139  
235030     NaN  26857-15420  
235031     NaN   26857-2009

## Check embeddings file

In [244]:
node_data_file_predict = os.path.join(PREDICT_DIR, "predict_top50_vs_all_node_data_embs_300.csv.gz")
node_data_file_train = os.path.join(content_api, "training_node_data_fixd.csv.gz")

In [ ]:
embeddings = pd.read_csv(node_data_file_predict, compression="gzip", index_col=0)

In [238]:
embeddings.shape

(206038, 513)

In [245]:
embeddings1 = pd.read_csv(node_data_file_train, compression="gzip", index_col=0)

In [249]:
c_all = set(list(embeddings.content_id) + list(embeddings1.content_id))
len(c_all) 

220941

In [250]:
content_ids_edges = set(list(merged_edges.src_cid)+list(merged_edges.dest_cid)) 
# content_ids_functedges = set(list(edges.src_cid)+list(edges.dest_cid)) 
len(content_ids_edges)

73173

In [251]:
len(content_ids_edges.difference(c_all))

0

### Count duplicate entries in embeddings

In [252]:
all_embeddings = pd.concat([embeddings, embeddings1], sort=True, ignore_index=True)

In [253]:
ids = all_embeddings['content_id']
all_embeddings[ids.isin(ids[ids.duplicated()])].sort_values("content_id")

0         1        10       100       101       102       103  \
3994   -0.016188  0.031590  0.012646 -0.021106 -0.000546 -0.053907 -0.013014   
221573 -0.016188  0.031590  0.012646 -0.021106 -0.000546 -0.053907 -0.013014   
259292 -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
266762 -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
230390 -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
6234   -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
249902 -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
259432 -0.055287 -0.020550  0.025845  0.049996 -0.049346 -0.053477  0.036689   
247853 -0.048413  0.046790 -0.057513 -0.045078  0.002419 -0.018005  0.054131   
10696  -0.048413  0.046790 -0.057513 -0.045078  0.002419 -0.018005  0.054131   
14454  -0.048617 -0.003174  0.064009 -0.056477  0.027464  0.047275  0.019767   
262643 -0.048617 -0.003174  0.064009 -0.056477  0.027464  0.047275  0.019767   
15591  -0.031045  0.037224 -0.058240 -0.038744 -0.002370 -0.050649  0.007794   
267292 -0.031045  0.037224 -0.058240 -0.038744 -0.002370 -0.050649  0.007794   
859     0.057785 -0.058796  0.010655 -0.007762  0.019277 -0.047553  0.013506   
209492  0.057785 -0.058796  0.010655 -0.007762  0.019277 -0.047553  0.013506   
15676  -0.059489 -0.030290 -0.037194 -0.051347 -0.008148 -0.007840  0.059874   
267559 -0.059489 -0.030290 -0.037194 -0.051347 -0.008148 -0.007840  0.059874   
266863  0.015428  0.004334 -0.053739 -0.008096  0.007777 -0.024524  0.058394   
15475   0.015428  0.004334 -0.053739 -0.008096  0.007777 -0.024524  0.058394   
237519 -0.011832  0.044025 -0.000041 -0.009414  0.033019 -0.000645  0.077963   
8021   -0.011832  0.044025 -0.000041 -0.009414  0.033019 -0.000645  0.077963   
227480 -0.058246 -0.043585  0.061405 -0.020520 -0.011086 -0.054329  0.047438   
5493   -0.058246 -0.043585  0.061405 -0.020520 -0.011086 -0.054329  0.047438   
229894 -0.055664 -0.016460  0.059309 -0.012764  0.004173 -0.050656  0.006683   
6115   -0.055664 -0.016460  0.059309 -0.012764  0.004173 -0.050656  0.006683   
12237  -0.016149 -0.029472  0.012153 -0.026216  0.017537 -0.045590  0.055533   
253861 -0.016149 -0.029472  0.012153 -0.026216  0.017537 -0.045590  0.055533   
16613  -0.039580 -0.007586 -0.052892 -0.036186  0.009042 -0.060673  0.042845   
271111 -0.039580 -0.007586 -0.052892 -0.036186  0.009042 -0.060673  0.042845   
...          ...       ...       ...       ...       ...       ...       ...   
8322    0.025430 -0.038182 -0.038882 -0.010376  0.047259 -0.046518  0.036413   
9525   -0.056468 -0.020103  0.017348 -0.021931 -0.039474 -0.011923  0.056007   
243067 -0.056468 -0.020103  0.017348 -0.021931 -0.039474 -0.011923  0.056007   
257629  0.009711 -0.001028  0.041777 -0.034499 -0.041695 -0.059442 -0.036608   
13212   0.009711 -0.001028  0.041777 -0.034499 -0.041695 -0.059442 -0.036608   
250270 -0.045597  0.053418  0.011221  0.040182  0.020054 -0.029931 -0.030434   
11307  -0.045597  0.053418  0.011221  0.040182  0.020054 -0.029931 -0.030434   
15704  -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
260293 -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
267678 -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
228672 -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
243865 -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
217608 -0.050235  0.044685 -0.012129  0.058120 -0.053051 -0.058634 -0.028795   
235523  0.009626 -0.052833  0.059169  0.029810 -0.037649 -0.024170 -0.030532   
240203  0.009626 -0.052833  0.059169  0.029810 -0.037649 -0.024170 -0.030532   
8742    0.009626 -0.052833  0.059169  0.029810 -0.037649 -0.024170 -0.030532   
237921 -0.037927  0.036834 -0.035781  0.032931 -0.054404 -0.018669  0.029044   
8122   -0.037927  0.036834 -0.035781  0.032931 -0.054404 -0.018669  0.029044   
15340  -0.04026

In [254]:
all_embeddings.drop_duplicates("content_id", inplace=True)

### Map `content_id` to `node_id` again

In [255]:
nid_cid = dict(zip(merged_edges.source, merged_edges.src_cid))
for nid,cid in zip(merged_edges.target, merged_edges.dest_cid):
    if nid not in nid_cid.keys():
        nid_cid[nid] = cid

In [259]:
cols = [str(i) for i in range(0,512)]
cols[-1]

'511'

In [260]:
embeddings_dict = dict(zip(all_embeddings.content_id, all_embeddings[cols].values))

In [265]:
all_embeddings[all_embeddings.content_id=="80f7a69c-5cdc-4bd5-a16c-15d669ba2cc5"]["3"]

0    0.008009
Name: 3, dtype: float64

In [266]:
rowlist = []
index = []
for nid,cid in nid_cid.items():
    row_dict = {}
    row_dict["content_id"] = cid
    for i,emb in enumerate(embeddings_dict[cid]):
        row_dict[i] = emb
    rowlist.append(row_dict)  
    index.append(nid)
new_embeddings = pd.DataFrame(rowlist, index=index)

In [267]:
new_embeddings.head()

content_id         0         1         2  \
33265  5ef7560d-7631-11e4-a3cb-005056011aef -0.047676 -0.027639  0.056349   
58314  1e333395-5dd5-4452-96a3-fbe939928761 -0.033119  0.049818  0.042104   
23720  d612c61e-22f4-4922-8bb2-b04b9202126e -0.067857 -0.015453 -0.028865   
25053  305acc88-488e-44ec-8e5a-1a5de7819ba9 -0.055095  0.003975  0.034079   
14044  305acc88-488e-44ec-8e5a-1a5de7819ba9 -0.055095  0.003975  0.034079   

              3         4         5         6         7         8    ...     \
33265 -0.009413 -0.020528 -0.045546  0.068116  0.018213  0.016371    ...      
58314 -0.017883 -0.007372 -0.054978  0.064178  0.062335  0.011622    ...      
23720  0.020038  0.017298 -0.064374  0.081447 -0.008263  0.021375    ...      
25053 -0.016298 -0.041664 -0.046474  0.056651  0.056367  0.050351    ...      
14044 -0.016298 -0.041664 -0.046474  0.056651  0.056367  0.050351    ...      

            502       503       504       505       506       507       508  \
33265  0.019222  0.066869  0.036356 -0.010981  0.030189 -0.045008  0.007033   
58314 -0.034333  0.064000  0.063919 -0.037945  0.054777  0.023922  0.002650   
23720  0.028531  0.019493  0.058379 -0.001819  0.070202 -0.067259  0.041385   
25053  0.022530  0.056369  0.056647 -0.035750  0.054273  0.048701  0.021803   
14044  0.022530  0.056369  0.056647 -0.035750  0.054273  0.048701  0.021803   

            509       510       511  
33265 -0.047787 -0.001289 -0.063644  
58314 -0.063124 -0.019809 -0.057015  
23720 -0.066728  0.001454 -0.014244  
25053 -0.055919  0.002437 -0.052509  
14044 -0.055919  0.002437 -0.052509  

[5 rows x 513 columns]

In [268]:
merged_edges[merged_edges.source==25053]

dest_cid  \
3       305acc88-488e-44ec-8e5a-1a5de7819ba9   
1087    32d4d2c8-2f9c-406f-96e4-65222074642a   
1216    686e6d1d-78c2-4c9d-b863-be5c5d8ed787   
1220    32d4d2c8-2f9c-406f-96e4-65222074642a   
1221    d6c2de5d-ef90-45d1-82d4-5f2438369eea   
2019    305acc88-488e-44ec-8e5a-1a5de7819ba9   
3593    305acc88-488e-44ec-8e5a-1a5de7819ba9   
3608    305acc88-488e-44ec-8e5a-1a5de7819ba9   
5181    b3fb8ff7-ac2e-4440-8c0f-227eb038c436   
6088    58b05bc2-fde5-4a0b-af73-8edc532674f8   
6095    305acc88-488e-44ec-8e5a-1a5de7819ba9   
8232    1e333395-5dd5-4452-96a3-fbe939928761   
8233    f3bbdec2-0e62-4520-a7fd-6ffd5d36e03a   
8240    5ef7560d-7631-11e4-a3cb-005056011aef   
8960    dc1a1744-4089-43b3-b2e3-4e397b6b15b1   
9412    e1067450-7d13-45ff-ada4-5e3dd4025fb7   
11047   4ab4764d-d9ce-425f-a8cc-aaba4a38be09   
12557   ea1f74cd-7c16-48d8-bf79-1a45b80508ef   
13971   305acc88-488e-44ec-8e5a-1a5de7819ba9   
14132   305acc88-488e-44ec-8e5a-1a5de7819ba9   
14654   76698ffe-70ab-4fda-be0d-755234f6d340   
14989   32d4d2c8-2f9c-406f-96e4-65222074642a   
16383   305acc88-488e-44ec-8e5a-1a5de7819ba9   
16393   f3bbd17c-48e2-403b-96ae-c09cb5d4dc35   
17754   df6fcdc4-62bc-4e05-90e3-cb2e452453a4   
18331   f9015c31-61c2-4504-8eb0-242cd75aee19   
22031   305acc88-488e-44ec-8e5a-1a5de7819ba9   
22335   34b852ce-9045-4431-9a16-6609d307cdb7   
24099   305acc88-488e-44ec-8e5a-1a5de7819ba9   
25523   764a2ce0-eae5-414a-a4a9-2217507165fb   
...                                      ...   
147937  07cfa36e-cf61-4358-a2b4-44518499d288   
148014  df86209b-1f0d-4280-b637-494ef61d0eac   
155899  87e2748f-2e9b-4681-8baa-778b6d326a8a   
156337  f9015c31-61c2-4504-8eb0-242cd75aee19   
159221  91504a10-f697-42a3-a779-f238b4955ea9   
164385  5b1cc445-222b-499c-b042-32cc84c8f52c   
165541  34b852ce-9045-4431-9a16-6609d307cdb7   
166675  34b852ce-9045-4431-9a16-6609d307cdb7   
170710  34b852ce-9045-4431-9a16-6609d307cdb7   
172961  6bcb87c7-0d26-4fb0-8079-c2dfa09220f3   
173599  34b852ce-9045-4431-9a16-6609d307cdb7   
173849  34b852ce-9045-4431-9a16-6609d307cdb7   
176494  089a887a-4d63-4da3-b46b-e81fb940dc0c   
177940  e1067450-7d13-45ff-ada4-5e3dd4025fb7   
179713  34b852ce-9045-4431-9a16-6609d307cdb7   
183276  5f560a94-7631-11e4-a3cb-005056011aef   
197638  686e6d1d-78c2-4c9d-b863-be5c5d8ed787   
198595  b3fb8ff7-ac2e-4440-8c0f-227eb038c436   
200045  f1b13904-a949-4a1f-b731-e87f1aa14daa   
202977  34b852ce-9045-4431-9a16-6609d307cdb7   
233900  5ee55afc-7631-11e4-a3cb-005056011aef   
233901  5ee54d1e-7631-11e4-a3cb-005056011aef   
233902  5f11c887-7631-11e4-a3cb-005056011aef   
233907  0b0ff1cc-1532-4cd3-836b-0b7ab0359a4b   
233909  5f1421df-7631-11e4-a3cb-005056011aef   
233910  fa7bb7fd-b4bc-45a2-8958-1f0bfa4c3e56   
233911  240281d1-4e92-428b-a784-9fa085c44777   
233912  5efadc60-7631-11e4-a3cb-005056011aef   
233914  5ee548bd-7631-11e4-a3cb-005056011aef   
233917  19786269-f057-405b-bea4-941b3b194ca4   

                                                dest_node  source  \
3                          /uk-family-visa/partner-spouse   25053   
1087                                    /settle-in-the-uk   25053   
1216                                       /family-permit   25053   
1220                    /settle-in-the-uk/y/none-of-these   25053   
1221                                   /government/brexit   25053   
2019                         /uk-family-visa/private-life   25053   
3593                  /uk-family-visa/provide-information   25053   
3608                 /uk-family-visa/knowledge-of-english   25053   
5181                  /healthcare-immigration-application   25053   
6088                                             /contact   25053   
6095                         /uk-family-visa/proof-income   25053   
8232                                           /visa-fees   25053   
8233                                                    /   25053   
8240    /government/publications/guidance-for-dependan...   25053   
8960                    

In [269]:
train_network = os.path.join(DATA_DIR,"train_network")
merged_file = os.path.join(train_network, "merged_edges_struct_funct.csv.gz")
new_embeddings_file = os.path.join(train_network, "merged_embs_struct_funct.csv.gz")

In [270]:
merged_edges.to_csv(merged_file, compression="gzip", index=False)
new_embeddings.to_csv(new_embeddings_file, compression="gzip")